# MetArea
MetArea tool proposes Meta Area under curve approach for predicting structural variability and cooperative binding of transcription factors in ChIP-seq data.

# Description
For the model of transcription factor (TF) binding sites motif recognition, the recognition accuracy is calculated as the partial area under ROC curve (pAUC, [Tsukanov et al., 2022](https://doi.org/10.3389/fpls.2022.938545)). The term ROC means Receiver Operating Characteristic curve. Generalizing the ROC curves and pAUC values of the two separate models to a single joint model reveals whether the joint performance of two motifs competes with those of the participating motifs. Testing multiple possible combinations of motifs allows to detect most strongly reinforcing each other motifs. The tested motifs can represent either structurally different types of binding sites for the same TF, or binding sites of different TFs acting as part of a single multi-protein complex. Thus, MetArea predicts motifs with synergistically related functions in gene transcription regulation. The synergy criterion for the pair of motifs 1 and 2 requires the higher pAUC1&2 value of the jount model 1&2 compared to the pAUC values of both participating motif models pAUC1 and pAUC2, Ratio = pAUC(1&2) / Max(pAUC1, pAUC2) > 1. MetArea consider ready ROC curves for TF binding site motif recognition models, thereby it can be applied to the motifs either of the traditional model of Position Weight Matrix (PWM) or an alternative model, e.g. [SiteGA](https://github.com/parthian-sterlet/sitega). 

# Algorithm
MetArea algorithm implies computation of the Receiver Operating Characteristic (ROC) curve for a combination of two motif models. For each motif model the ROC curve represents the relationship between the rates of true positive (TP) and false positive (FP) predictions. For each recognition threshold, the TP rate denotes the fraction of nucleotide sequences from the foreground set (ChIP-seq peaks) containing predicted sites of a motif model. For each recognition threshold, the FP rate is defined as the frequency of the motif in the background set of nucleotide sequences ([Tsukanov et al., 2022](https://doi.org/10.3389/fpls.2022.938545)). The background set means genomic sequences with A/T content matching that for the sequences from the foreground set, details in [AntiNoise](https://denovosea.icgbio.ru/antinoise/). Preliminary, for each motif model, the promoter sequences of protein coding genes from whole genomes are used to compute -Log10(ERR) values (Expected Recognition Rate, ERR). These ERR values measure various motif models in the uniform scale ([Levitsky et al., 2019](https://doi.org/10.1093/nar/gkz800); [Tsukanov et al., 2022](https://doi.org/10.3389/fpls.2022.938545)). Thus, in terms of the FP rate computed through ERR values, for any pair of sites of two models we can estimate which of two sites respects higher affinity. 

For a combination of two motif models, the joint TP rate is determined in the same way as for single motifs. The joint TP rate is the fraction of sequences containing predicted sites of at least of one of two motif models. The joint FP rate certainly requires the correction since predicted sites of two motif models may overlap each other in a nucleotide sequence. In particular, following the methodology of [MCOT](https://doi.org/10.1093/nar/gkz800) only the following three different cases are possible for certain two sites of two distinct motif models: 
- two sites are located with a spacer since they do not overlap;
- two sites have the full overlap, i.e. one of them is located entirely inside the other;
- two sites have only a partial overlap, this means that when two sites overlap, neither site can be completely located within the other site.

Correspondingly, MetArea proposes the following approach to correct the joint FP rate for the joint motif model:
- if two sites of two models do not overlap, then the joint FP rate increases additively by the impact of two, dFP = 2, two distinct sites in a sequence are detected;
- if these two sites possess the full overlap, then the joint FP rate increases only by the impact of one, dFP = 1, only one site in a sequence is detected;
- if two sites show the partial overlap, then the joint FP rate increases the non-integer impact between one and two (1 < dFP < 2), depending on the ratios between the overlap length (L) of two motifs, and the lengths of motifs W1 and W2, dFP = (L/W1 + L/W2)/2 = L*(W1 + W2)/(W1*W2)/2.

The cases of full or partial overlaps imply that joined FP rate possesses the best estimate of ERR for two sites, -Log10(ERR1&2) = Max( -Log10(ERR1),  -Log10(ERR2)). 

To differentiate significantly similar pairs of motifs among analyzed pairs of PWM motifs, for all pairs of PWM motifs the significance of similarity is estimated by the motif similarity approach from [MCOT, (Levitsky et al., 2019)](https://doi.org/10.1093/nar/gkz800)

# Input data

- FASTA file for the set of foreground sequences (peaks). This set represents top-scored peaks derived from ChIP-seq experiment (1000 peaks of the best quality is the default recomended option).
- FASTA file for the set background sequences. This set consisted of sequences extracted from the whole genome for the respective species, use [AntiNoise](https://denovosea.icgbio.ru/antinoise/) with the default either genomic or promoters options. These options mean extraction of background sequences from the whole genome or only from promoter regions of all protein coding genes. It is expected the promoter option is the best one. Do not use the shuffled sequences or any other sequences generated by Markov models of different orders as the background set, this drastically reduces the quality of the output results.
- Two distinct motif models, e.g. Position Frequency Matrices (PFM) representing PWM motifs.
- Two lists of recognition thresholds for two motif models and the respective ERR values. These lists are computated as the recognition score distributions for the set of whole-genome promoter sequences selected for the respective species, as it was described earlier for [MCOT](https://webmcot.sysbio.cytogen.ru/) [(Levitsky et al., 2019)](https://doi.org/10.1093/nar/gkz800) and [SiteGA](https://github.com/parthian-sterlet/sitega) [(Tsukanov et al., 2022)](https://doi.org/10.3389/fpls.2022.938545).

# Output data

- list of performanes pAUC for all motifs
- list of all pairs of motifs with synergy marks, Ratio > 1 or Ratio < 1
- esiimation of motifs similarities for all pairs of PWM motif models

# Running modes

- One PWM motif model vs. another PWM motif model.
- One traditional motif model PWM vs. one alternative motif model [SiteGA](https://github.com/parthian-sterlet/sitega). The SiteGA model was chosen as the most methodologically different from the tradiotional PWM.
- One PWM motif model vs. all motifs for known TFs from a given collection from [Hocomoco v12](http://hocomoco12.autosome.ru/) or [JASPAR2024](https://jaspar.elixir.no/).
- Best performed motifs from the collection of PWM motifs for known TFs for certain taxon or species. The performances of all motifs from a given collection are estimated by pAUC, then all possible {N * (N - 1)/2} pairwise combinations of motifs for the top-scored M motifs of all N motifs in the collection. Available options includes the collections of 1420/1142 motifs for 942/713 human/mouse TFs from [Hocomoco v12](http://hocomoco12.autosome.ru/) ([Vorontsov et al., 2024](https://doi.org/10.1093/nar/gkad1077)), of 556/151 motifs for 555/148 plant/insect TFs from [JASPAR2024](https://jaspar.elixir.no/) [Rauluseviciute et al., 2024](https://doi.org/10.1093/nar/gkad1059).

# Source code and command line arguments

- One PWM motif model vs. another PWM motif model, [pauc_forback_2motifs_only.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_2motifs_only.cpp)
1. input FASTA file,foreground set pf sequences, peaks, example - [top 1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
2. input FASTA file, background set of sequences, genomic sequences, [A/T-matched random pieces of DNA of the same length from promoters](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/) 
3. input binary files for the first motif model (PWM1), [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.0.P.B.binary). This file integrates in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.0.P.B.pcm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.0.P.B.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.0.P.B.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel).
4. input binary files for the second motif model (PWM2), [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.1.SM.B.binary). This file integrates in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.1.SM.B.pcm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.1.SM.B.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.1.SM.B.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel).
5. output file, pAUC values for the motifs models PWM1 and PWM2 and for the joint model PWM1&PWM2, [example, pAUC values for PWM1, PWM2 and joint PWM1&PWM2 models](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.0.P.B_BHA15.H12CORE.1.SM.B.pauc), the significance of similarity between two motifs
6. output file, ROC curve for the motif model PWM1
7. output file, ROC curve for the motif model PWM2
8. output file, ROC curve for the joint model PWM1&PWM2, [example, ROC curve for joint PWM1&PWM2 model](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.0.P.B_BHA15.H12CORE.1.SM.B.roc)
   
- One traditional motif model PWM vs. one alternative motif model [SiteGA](https://github.com/parthian-sterlet/sitega), [pauc_forback_pwm_sga_only.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_pwm_sga_only.cpp)
1. input FASTA file,foreground set pf sequences, peaks, example - [top 1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
2. input FASTA file, background set of sequences, genomic sequences, [A/T-matched random pieces of DNA of the same length from promoters](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/) 
3. input binary file for PWM motif model, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_pwm.binary). This file integrates in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel). 
4. input binary file for SiteGA motif model, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_sga.binary). This file integrates in the binary format [the matrix of SiteGA model](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_sga.mat), including the motif length, [the total number of recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_sga_ERR.dist). The binary files for SiteGA models are generated by [SiteGA thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [SiteGA](https://github.com/parthian-sterlet/sitega)
5. output file, pAUC values for the motif models PWM and SiteGA and for the joint model PWM&SiteGA, [example, pAUC values for PWM, SiteGA and joint PWM&SiteGA models](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_pwm_sga.pauc)
6. output file, ROC curve for the motif model PWM
7. output file, ROC curve for the motif model SiteGA
8. output file, ROC curve for the joint model PWM&SiteGA, [example, ROC curve PWM&SiteGA model](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_pwm_sga.roc)

- One PWM motif model vs. all motifs for known TFs from a given collection, [pauc_forback_anc_lib.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_anc_lib.cpp)
1. input FASTA file,foreground set pf sequences, peaks, example - [top 1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
2. input FASTA file, background set of sequences, genomic sequences, [A/T-matched random pieces of DNA of the same length from promoters](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/)
3. input binary file for PWM motif model, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_pwm.binary). This file integrates in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel).
4. input binary file for the collection of PWM motif models, available options: two collections of human and murine motifs from [Hocomoco v12](http://hocomoco12.autosome.ru/), see archive files in [partners folder](https://github.com/parthian-sterlet/metarea/tree/main/partners), two collections for plant and insect TFs from [JASPAR2024](https://jaspar.elixir.no/), [JASPAR2024_insects](https://github.com/parthian-sterlet/metarea/blob/main/partners/jaspar24_dm6.binary) and [JASPAR2024_plants](https://github.com/parthian-sterlet/metarea/blob/main/partners/jaspar24_dm6.binary) These files in the binary format integrate for all motifs in each collection [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel).
5. input binary file for the pairwise similarity between the motifs from the collection of PWM motifs, examples for the human/mouse collections from [Hocomoco v12](http://hocomoco12.autosome.ru/), and for the plants/insects collections from [JASPAR2024](https://jaspar.elixir.no/) are provided in [partners folder](https://github.com/parthian-sterlet/metarea/tree/main/partners), e.g. for [data for insects motifs](https://github.com/parthian-sterlet/metarea/blob/main/partners/jaspar24_dm6_sim.binary)
6. output file, list of joint pAUC values satisfying the synergy criterion Ratio > 1, [example list of joint pAUC values for synergetic pairs of motifs](https://github.com/parthian-sterlet/metarea/blob/main/examples/anc_lib/PEAKS039234_BHLHA15_pwm.pauc), and the corresponding list of the significance of similarity of motifs in pairs
7. output file, list of joint pAUC values for all pairs of motifs, [example list of joint pAUC values for all pairs of motifs](https://github.com/parthian-sterlet/metarea/blob/main/examples/anc_lib/PEAKS039234_BHLHA15_pwm.log) including the corresponding list of the significance of similarity of motifs in pairs

- Best performed motifs from the collection of PWM motifs for known TFs for certain taxon or species, [pauc_forback_2motifs.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_2motifs.cpp)
1. input FASTA file,foreground set pf sequences, peaks, example - [top 1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
2. input FASTA file, background set of sequences, genomic sequences, [A/T-matched random pieces of DNA of the same length from promoters](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/)
3. input binary file for the collection of PWM motif models, available options: two collections of human and murine motifs from [Hocomoco v12](http://hocomoco12.autosome.ru/), see archive files in [partners folder](https://github.com/parthian-sterlet/metarea/tree/main/partners), two collections for plant and insect TFs from [JASPAR2024](https://jaspar.elixir.no/), [JASPAR2024_insects](https://github.com/parthian-sterlet/metarea/blob/main/partners/jaspar24_dm6.binary) and [JASPAR2024_plants](https://github.com/parthian-sterlet/metarea/blob/main/partners/jaspar24_dm6.binary) These files in the binary format integrate for all motifs in each collection [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel)
4. input binary file for the pairwise similarity between the motifs from the collection of PWM motifs, examples for the human/mouse collections from [Hocomoco v12](http://hocomoco12.autosome.ru/), and for the plants/insects collections from [JASPAR2024](https://jaspar.elixir.no/) are provided in [partners folder](https://github.com/parthian-sterlet/metarea/tree/main/partners), e.g. for [data for insects motifs](https://github.com/parthian-sterlet/metarea/blob/main/partners/jaspar24_dm6_sim.binary)
5. int value, the number of M top-scored motifs to check pairwise synergy between motifs
6. output file, the matrix of all (M*(M-1)) joint pAUC values, all (M*(M-1)) pairwise Ratio values, and all (M*(M-1)) significance of pairwise similarity for all tested pairs formed by the selected M top-scored motifs, [example_of output_matrices](https://github.com/parthian-sterlet/metarea/blob/main/examples/lib_lib/PEAKS039234_BHLHA15_Q9QYC3_MACS2.auc_mat)
7. output file pAUC, list of joint pAUC values for all pairs of motifs formed by top-scored M motifs, these pairs satisfy the synergy criterion Ratio > 1, [example list of joint pAUC values for synergetic pairs of motifs](https://github.com/parthian-sterlet/metarea/blob/main/examples/lib_lib/PEAKS039234_BHLHA15_Q9QYC3_MACS2.auc_mat) including the corresponding list of the significance of similarity of motifs in pairs
8. output file log1, list of pAUC values for all single motifs from the collection, [example list of pAUC values fo single motifs](https://github.com/parthian-sterlet/metarea/blob/main/examples/lib_lib/PEAKS039234_BHLHA15_Q9QYC3_MACS2.auc_log1)
9. output file log2, list of joint pAUC values for all pairs of motifs formed by top-scored M motifs, [example list of joint pAUC values for all pairs of motifs](https://github.com/parthian-sterlet/metarea/blob/main/examples/lib_lib/PEAKS039234_BHLHA15_Q9QYC3_MACS2.auc_log2) including the corresponding list of the significance of similarity of motifs in pairs

# Command line examples

- One PWM motif model vs. another PWM motif model, [pauc_forback_2motifs_only.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_2motifs_only.cpp)

[pwm_sga](https://github.com/parthian-sterlet/metarea/blob/main/run/pwm_sga) command line for PWM and SiteGA motif models for [ChIP-seq peaks dataset PEAKS039234 for BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa), [GSE86289](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE86289), pancreas of adult mice. Two motif models derived from this dataset with _de novo_ motif search.

- One traditional motif model PWM vs. one alternative motif model [SiteGA](https://github.com/parthian-sterlet/sitega), [pauc_forback_pwm_sga_only.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_pwm_sga_only.cpp)
 
[pwm_pwm](https://github.com/parthian-sterlet/metarea/blob/main/run/pwm_pwm) command line for two PWM models [BHA15.H12CORE.0.P.B.pcm](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.0.P.B.pcm) and [BHA15.H12CORE.1.SM.B.pcm](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_pwm/BHA15.H12CORE.1.SM.B.pcm) respecting two structurally distinct motif types for [murine BHLHA15 TF](https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:22265)  from [Hocomoco v12](http://hocomoco12.autosome.ru/) and [ChIP-seq peaks dataset PEAKS039234 for BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa)

- One PWM motif model vs. all motifs for known TFs from a given collection, [pauc_forback_anc_lib.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_anc_lib.cpp)
 
[anc_lib](https://github.com/parthian-sterlet/metarea/blob/main/run/anc_lib) command line for PWM motif model derived by _de novo_ motif search for [ChIP-seq peaks dataset PEAKS039234 for BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) and the collection of motifs for murine TFs from [Hocomoco v12](http://hocomoco12.autosome.ru/)

- Best performed motifs from the collection of PWM motifs for known TFs for certain taxon or species, [pauc_forback_2motifs.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_2motifs.cpp)

[lib_lib](https://github.com/parthian-sterlet/metarea/blob/main/run/anc_lib) command line to test [ChIP-seq peaks dataset PEAKS039234 for BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) for top-scoring pairwise combiantaions of motifs from the [Hocomoco v12](http://hocomoco12.autosome.ru/) collection for known murine TFs.
