# MetArea
MetArea tool proposes Meta Area under curve computation to estimate cooperative transcription factor binding in ChIP-seq data.

# Description
For a pairs of motifs conventional recognition accuracy is calculated as a partial area under curve (pAUC, [Tsukanov et al., 2022](https://doi.org/10.3389/fpls.2022.938545)). Generalisation of this accuracy measure to joint curve for these two motifs allows to compute a common accuracy for these two motifs. Testing of multiple possible combinations of motifs allows to detect most strongly reinforcing each other. The tested motifs can present either structurally different types of binding sites for the same transcription factor, or binding sites of diferent transcription factors acting as part of a single multi-protein complex. Thus, MetArea predicts motifs with synergistically related functions. Synergy is recognizwd as higher jount pAUC value of a combination of two model compared to any pAUC value from single models, Ratio = pAUC(1&2) / Max(pAUC1, pAUC2) > 1.
MetArea can be uniformly applied to the motifs either of the traditional (Position Weight Matrix) or alternative models, e.g. [SiteGA](https://github.com/parthian-sterlet/sitega). 

# Algorithm
MetArea algorighm implies computation of the Receiver operating characteristic (ROC) curve for a combination of two motifs. For a single motif the ROC curve represents the relationship between the rates of true positive (TP) and false positive (FP) presictions. For each recognition threshold TP rate denotes the fraction of sequences from the foreground set (ChIP-seq peaks) containing predicted sites of a motif. For each recognition threshold we define FP rate as the freqiency of a motif in a backgound set of seqiences ([Tsukanov et al., 2022](https://doi.org/10.3389/fpls.2022.938545)). The background set means genomic sequences with A/T content matching that for the sequences from the foregdound set. Preliminary, for each motif model, we use the promoters sequences of pretein coding genes from whole genomes to compute -Log10(ERR) values (Expected Recognition Rate, ERR). These values measure various motif model in the uniform scale. Thus, in terms of FP rate computed through ERR, for any pair of sites of two models we can estimate which of two sites respects higher affinity. 

For a combination of two motifs, the joint TP rate is defined similarly as ones for single motifs. The joint TP rate is a fraction of sequences containing hits of at least of one of two motifs. The joint FP rate requres correction since sites of two models may overlap each other. 

- If two sites of two models do not overlap, the joint FP rate increases additively by two, dFP = 2. We see two distinct sites in a sequence.
- If these two sites possess the full overlap, the joint FP rate increases only by one, dFP = 1. We see one site in a sequence.
- If two sites show the partial overlap, the joint FP rate increases an non-integer impact between one and two (1 < dFP < 2), depending on the ratios between the overlap length (OL), and the lengths of motifs W1 and W2, dFP = 0.5 * (OL / W1 + OL / W2) = 0.5 * OL * (W1 + W2) / (W1 * W2)).

The cases of full or partial overlaps imply that joined FP rate possesses the best estimate of ERR for two sites, -Log10(ERR,joined) = Max( -Log10(ERR1),  -Log10(ERR2)).

# Main input data
- fasta files for the sets of foreground (peaks) and background sequences. The former set respects to top-scored peaks derived from ChIP-seq experiment (1000 peaks of best quality is default option), the latter set consisted of genomic sequences extracted from the whole genome for the respective species, use [AntiNoise](https://denovosea.icgbio.ru/antinoise/) with default genomic or promoters options. It is expected the promoters option is the best one. Do not use shuffled sequences or any other sequences generated by Markov models of different orders as a background set - such an option drastically reduces the quality of the output results.
- a couple of motif models, e.g. a Position Frequency Matrices representing PWM motifs
- a full set of model recognition thresholds through computation od recognitionin score distribution for the set of whole-genome promoter sequences selected for respective species, as it was described earlier for [MCOT](https://webmcot.sysbio.cytogen.ru/) [(Levitsky et al., 2019)](https://doi.org/10.1093/nar/gkz800) and [SiteGA](https://github.com/parthian-sterlet/sitega) [(Tsukanov et al., 2022)](https://doi.org/10.3389/fpls.2022.938545). 

# Running modes
1. Test of a couple of motif models, both models are represented by traditional PWMs
2. Test of a full collection of motifs for known TFs for certain taxon or species, the performaces of all motifs from a given collection is estimated by pAUC, then all possible {N * (N - 1)/2} pairwise combinations of motifs for the top-scored N motifs of all M motifs in a collection. Available options includes the collections of 1420/1142 motifs for 942/713 human/mouse TFs from [Hocomoco version 12](http://hocomoco12.autosome.ru/) ([Vorontsov et al., 2024](https://doi.org/10.1093/nar/gkad1077)), of 556/151 motifs for 555/148 plant/insect TFs from [JASPAR2024](https://jaspar.elixir.no/) [Rauluseviciute et al., 2024](https://doi.org/10.1093/nar/gkad1059)
3. Test of all possible M pairwise combinations of motifs for given motif and all motifs from a library
4. Test of a couple of motif models, the first and second models are represented by traditional PWM and the alternative [SiteGA](https://github.com/parthian-sterlet/sitega). The latter was chosen as the most methodologically different from the tradiotional one.

# Source code and command line arguments
Running mode #4, [pauc_forback_pwm_sga_only.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_pwm_sga_only.cpp)
1. path input fasta files (arguments #2, #3)
2. foreground set of sequences, fasta format, peaks, example - [top1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
3. background set of sequences, fasta format, genomic sequences, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters of protein coding genes extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/) 
4. pwm binary file, PFM, PWM, motif length, total number of thresholds, list of thresholds and ERR values, [PWM_binary](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm.binary), generated by 
[PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel)
5. sga binary file,  SiteGA weights, motif length, total number of thresholds, list of thresholds and ERR values, generated by 
[SiteGA thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [SiteGA](https://github.com/parthian-sterlet/sitega)
6. output file, ROC curves for single models PWM and SiteGA, joint ROC curve PWM&SiteGA
7. output file pAUC, list of three pAUC values
8. output file log, list of three pAUC values

File [run_pwm_sga](https://github.com/parthian-sterlet/metarea/blob/main/src/run_pwm_sga) shows the example command line implementing the combination of PWM and SiteGA motif models for [the peaks dataset PEAKS039234 for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
