# MetArea
MetArea tool proposes Meta Area under curve computation to estimate cooperative transcription factor binding in ChIP-seq data.

# Description
For a pairs of motifs conventional recognition accuracy is calculated as a partial area under curve (pAUC, [Tsukanov et al., 2022](https://doi.org/10.3389/fpls.2022.938545)). Generalisation of this accuracy measure to joint curve for these two motifs allows to compute a common accuracy for these two motifs. Testing of multiple possible combinations of motifs allows to detect most strongly reinforcing each other. The tested motifs can present either structurally different types of binding sites for the same transcription factor, or binding sites of diferent transcription factors acting as part of a single multi-protein complex. Thus, MetArea predicts motifs with synergistically related functions. Synergy is recognizwd as higher jount pAUC value pAUC(1&2) of a combination of two motif models compared to any pAUC value from single motif models pAUC1 and pAUC2, Ratio = pAUC(1&2) / Max(pAUC1, pAUC2) > 1. MetArea can be applied to the motifs either of the traditional PWM (Position Weight Matrix) or alternative models, e.g. [SiteGA](https://github.com/parthian-sterlet/sitega). 

# Algorithm
MetArea algorighm implies computation of the Receiver Operating Characteristic (ROC) curve for a combination of two motifs. For a single motif the ROC curve represents the relationship between the rates of true positive (TP) and false positive (FP) presictions. For each recognition threshold TP rate denotes the fraction of sequences from the foreground set (ChIP-seq peaks) containing predicted sites of a motif. For each recognition threshold we define FP rate as the freqiency of a motif in a backgound set of seqiences ([Tsukanov et al., 2022](https://doi.org/10.3389/fpls.2022.938545)). The background set means genomic sequences with A/T content matching that for the sequences from the foregdound set. Preliminary, for each motif model, we use the promoter sequences of protein coding genes from whole genomes to compute -Log10(ERR) values (Expected Recognition Rate, ERR). These values measure various motif model in the uniform scale. Thus, in terms of FP rate computed through ERR, for any pair of sites of two models we can estimate which of two sites respects higher affinity. 

For a combination of two motifs, the joint TP rate is defined similarly as ones for single motifs. The joint TP rate is a fraction of sequences containing hits of at least of one of two motifs. The joint FP rate requres correction since sites of two models may overlap each other. 

- If two sites of two models do not overlap, the joint FP rate increases additively by two, dFP = 2. We see two distinct sites in a sequence.
- If these two sites possess the full overlap, the joint FP rate increases only by one, dFP = 1. We see one site in a sequence.
- If two sites show the partial overlap, the joint FP rate increases an non-integer impact between one and two (1 < dFP < 2), depending on the ratios between the overlap length (L), and the lengths of motifs W1 and W2, dFP = (L/W1 + L/W2)/2 = L*(W1 + W2)/(W1*W2)/2.

The cases of full or partial overlaps imply that joined FP rate possesses the best estimate of ERR for two sites, -Log10(ERR,joined) = Max( -Log10(ERR1),  -Log10(ERR2)).

# Main input data
- FASTA file for the set of foreground sequences (peaks). This set represents top-scored peaks derived from ChIP-seq experiment (1000 peaks of best quality is default option).
- FASTA file for the set background sequences. This set consisted of sequences extracted from the whole genome for the respective species, use [AntiNoise](https://denovosea.icgbio.ru/antinoise/) with default genomic or promoters options. These options mean extraction of background sequences from a whole genome or only those of them mapped to promoters of protein coding genes. It is expected the promoters option is the best one. Do not use shuffled sequences or any other sequences generated by Markov models of different orders as a background set - such an option drastically reduces the quality of the output results.
- A couple of motif models, e.g. a Position Frequency Matrices representing PWM motifs
- Two lists of recognition thresholds for two motif models. These lists are computated as recognition score distribution for the set of whole-genome promoter sequences selected for respective species, as it was described earlier for [MCOT](https://webmcot.sysbio.cytogen.ru/) [(Levitsky et al., 2019)](https://doi.org/10.1093/nar/gkz800) and [SiteGA](https://github.com/parthian-sterlet/sitega) [(Tsukanov et al., 2022)](https://doi.org/10.3389/fpls.2022.938545). 

# Running modes
1. One PWM motif model vs. another PWM motif model.
2. Traditional PWM motif model vs. alternative [SiteGA](https://github.com/parthian-sterlet/sitega) motif model. The SiteGA model was chosen as the most methodologically different from the tradiotional PWM.
3. One PWM motif model vs. all motifs for known TFs from a given collection [Hocomoco version 12](http://hocomoco12.autosome.ru/) or [JASPAR2024](https://jaspar.elixir.no/).
4. Best performed motifs from the collection of PWM motifs for known TFs for certain taxon or species. The performances of all motifs from a given collection are estimated by pAUC, then all possible {N * (N - 1)/2} pairwise combinations of motifs for the top-scored M motifs of all N motifs in the collection. Available options includes the collections of 1420/1142 motifs for 942/713 human/mouse TFs from [Hocomoco version 12](http://hocomoco12.autosome.ru/) ([Vorontsov et al., 2024](https://doi.org/10.1093/nar/gkad1077)), of 556/151 motifs for 555/148 plant/insect TFs from [JASPAR2024](https://jaspar.elixir.no/) [Rauluseviciute et al., 2024](https://doi.org/10.1093/nar/gkad1059).

# Source code and command line arguments
One PWM motif model vs. another PWM motif model, [pauc_forback_2motifs_only.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_2motifs_only.cpp)
1. input FASTA file,foreground set pf sequences, peaks, example - [top1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
2. input FASTA file, background set of sequences, genomic sequences, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/) 
3,4 input binary files for two PWM motif models, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm.binary). This file integrates in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel). 
5. output file, ROC curves for single models PWM and SiteGA, the joint ROC curve (PWM & SiteGA) combiantion, see [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga.roc)
6. output file pAUC, list of three pAUC values
7. output file log, list of three pAUC values
   
Traditional PWM vs. alternative SiteGA motif models, [pauc_forback_pwm_sga_only.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_pwm_sga_only.cpp)
1. input FASTA file,foreground set pf sequences, peaks, example - [top1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
2. input FASTA file, background set of sequences, genomic sequences, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/) 
3. input binary file for PWM motif model, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm.binary). This file integrates in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel). 
4. input binary file for SiteGA motif model, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_sga.binary). This file integrates in the binary format [the matrix of SiteGA model](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_sga.mat), including the motif length, [the total number of recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_sga_ERR.dist). The binary files for SiteGA models are generated by [SiteGA thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [SiteGA](https://github.com/parthian-sterlet/sitega)
5. output file, ROC curves for single models PWM and SiteGA, the joint ROC curve (PWM & SiteGA) combiantion, see [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/pwm_sga.roc)
6. output file pAUC, list of three pAUC values
7. output file log, list of three pAUC values

One PWM motif model vs. all motifs for known TFs from a given collection, [pauc_forback_anc_lib.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_anc_lib.cpp)
1. input FASTA file,foreground set pf sequences, peaks, example - [top1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
2. input FASTA file, background set of sequences, genomic sequences, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/)
3. input binary file for PWM motif model, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm.binary). This file integrates in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel).
4. input binary file for the collection of PWM motif models, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm.binary). This file integrates in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel). 
5. output file pAUC, list of pAUC values satisfying the synergy criterion Ratio > 1
6. output file log, list of pAUC values for all pairs of motifs

Best performed motifs from the collection of PWM motifs for known TFs for certain taxon or species, [pauc_forback_2motifs.cpp](https://github.com/parthian-sterlet/metarea/blob/main/src/pauc_forback_2motifs.cpp)
1. input FASTA file,foreground set pf sequences, peaks, example - [top1000 peaks for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
2. input FASTA file, background set of sequences, genomic sequences, [example](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2_pb.fa) genomic sequences from promoters extracted by [AntiNoise](https://denovosea.icgbio.ru/antinoise/)
3. input binary file for the collection of PWM motif models, available options [JASPAR2024_insects](https://github.com/parthian-sterlet/metarea/blob/main/partners/jaspar24_dm6.binary), [JASPAR2024_plants](https://github.com/parthian-sterlet/metarea/blob/main/partners/jaspar24_dm6.binary), and Hocomoco12 collections for human amd mouse in archive files in [partners folder](https://github.com/parthian-sterlet/metarea/tree/main/partners). These files integrate in the binary format [PFM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pfm), [PWM](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15.pwm), [the motif length, and total number of  recognition thresholds, the full list of thresholds and ERR values](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_pwm_ERR.dist). The binary files for PWM models are generated by [PWM thresholds selection program](https://github.com/parthian-sterlet/mcot-kernel/blob/master/src/pwm_thr_err/pwm_iz_pwm_thr_dist0.cpp) from [MCOT](https://github.com/parthian-sterlet/mcot-kernel). 
4. int value, the number of M top-scored motifs to check pairwise synergy
5. output file pAUC, list of joint pAUC values for all pairs of motifs formed by top-scored M motifs, these pairs satisfy the synergy criterion Ratio > 1
6. output file log1, list of pAUC values for all motifs from the collection
7. output file log2, list of pAUC values for all pairs of motifs formed by top-scored M motifs

File [run_pwm_sga](https://github.com/parthian-sterlet/metarea/blob/main/src/run_pwm_sga) shows the example command line implementing the combination of PWM and SiteGA motif models for [the peaks dataset PEAKS039234 for mouse BHLHA15 TF](https://github.com/parthian-sterlet/metarea/blob/main/examples/PEAKS039234_BHLHA15_Q9QYC3_MACS2.fa) from [GTRD](https://gtrd.biouml.org/#!)
