# MetArea
MetArea means meta area under curve computation for cooperative transcription factor binding in ChIP-seq data.

For a pairs of motifs conventional recognition accuracy are estimated calculated as the values of partial area under curve (pAUC). Generalisation of this accuracy measure to joint curve for these two motifs allows to compute a common accuracy for these motifs. Testing of multiple possible combinations of motifs allows to detect most strongly reinforcing each other. The tested motifs can present either structurally different types of binding sites for the same transcription factor, or binding sites of diferent transcription factors acting as part of a single multi-protein complex. Thus, MetArea predicts motifs with synergistically related functions. 
MetArea is uniformly applied to the motif of the traditional (Position Weight Matrix) and alternative models (e.g. [SiteGA](https://github.com/parthian-sterlet/sitega). 
To run the tool it required 
- fasta files for the sets of foreground (peaks) and background sequences. The former set respects to top-scored peaks derived from ChIP-seq experiment (1000 peaks of best quality is default option), the latter set consisted of genomic sequences extracted from the whole genome for the respective species, use [AntiNoise](https://denovosea.icgbio.ru/antinoise/) with default genomic or promoters options.
- a couple of motif models, e.g. a Position Frequency Matrices representing PWM motifs
- a full set of model recognition thresholds through computation od recognitionin score distribution for the set of whole-genome promoter sequences selected for respective species, as it was described earlier for [MCOT](https://webmcot.sysbio.cytogen.ru/) [(Levitsky et al., 2019)](https://doi.org/10.1093/nar/gkz800) and [SiteGA](https://github.com/parthian-sterlet/sitega) [(Tsukanov et al., 2022)](https://doi.org/10.3389/fpls.2022.938545). For each threshold the -Log10(ERR) value (Expected Recognition Rate, ERR) means a correspomding freqiency of a motif. 

Tool has several running modes:
1. Test of a couple of motif models, both models are represented by traditional PWMs
2. Test of a full collection of motifs for known TFs for certain taxon or species, the performaces of all motifs from a given collection is estimated by pAUC, then all possible {N * (N - 1)/2} pairwise combinations of motifs for the top-scored N motifs of all M motifs in a collection. Available options includes the collections of 1420/1142 motifs for 942/713 human/mouse TFs from [Hocomoco version 12](http://hocomoco12.autosome.ru/) ([Vorontsov et al., 2024](https://doi.org/10.1093/nar/gkad1077)), of 556/151 motifs for 555/148 plant/insect TFs from [JASPAR2024](https://jaspar.elixir.no/) [Rauluseviciute et al., 2024](https://doi.org/10.1093/nar/gkad1059)
3. Test of all possible M pairwise combinations of motifs for given motif and all motifs from a library
4. Test of a couple of motif models, the first and second models are represented by traditional PWM amd alternative [SiteGA](https://github.com/parthian-sterlet/sitega)
